clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java

```diff
@@ -0,0 +1,265 @@
+
+    private final int heartbeatInterval = 1000;
+    private final long retryBackoffMaxMs = 3000;
+    private final long retryBackoffMs = 100;
+    private final String groupId = "group-id";
+
+    private Time mockTime;
+    private LogContext mockLogContext;
+    private CoordinatorRequestManager mockCoordinatorRequestManager;
+    private SubscriptionState mockSubscriptionState;
+    private HeartbeatRequestManager heartbeatRequestManager;
+    private MembershipManager mockMembershipManager;
+    private HeartbeatRequestManager.HeartbeatRequestState heartbeatRequestState;
+    private ConsumerConfig config;
+
+    private String memberId = "member-id";
+    private int memberEpoch = 1;
+    private ConsumerGroupHeartbeatResponseData.Assignment memberAssignment = mockAssignment();
+    private ErrorEventHandler errorEventHandler;
+
+    private ConsumerGroupHeartbeatResponseData.Assignment mockAssignment() {
+        return new ConsumerGroupHeartbeatResponseData.Assignment()
+            .setAssignedTopicPartitions(Arrays.asList(
+                new ConsumerGroupHeartbeatResponseData.TopicPartitions()
+                    .setTopicId(Uuid.randomUuid())
+                    .setPartitions(Arrays.asList(0, 1, 2)),
+                new ConsumerGroupHeartbeatResponseData.TopicPartitions()
+                    .setTopicId(Uuid.randomUuid())
+                    .setPartitions(Arrays.asList(3, 4, 5))
+            ));
+    }
+
+    @BeforeEach
+    public void setUp() {
+        mockTime = new MockTime();
+        mockLogContext = new LogContext();
+        Properties properties = new Properties();
+        properties.put(BOOTSTRAP_SERVERS_CONFIG, "localhost:9999");
+        properties.put(KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class);
+        properties.put(VALUE_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class);
+        properties.put(RETRY_BACKOFF_MS_CONFIG, "100");
+        properties.put(HEARTBEAT_INTERVAL_MS_CONFIG, heartbeatInterval);
+        config = new ConsumerConfig(properties);
+        mockCoordinatorRequestManager = mock(CoordinatorRequestManager.class);
+        when(mockCoordinatorRequestManager.coordinator()).thenReturn(Optional.of(new Node(1, "localhost", 9999)));
+        mockSubscriptionState = mock(SubscriptionState.class);
+        mockMembershipManager = spy(new MembershipManagerImpl(groupId));
+        heartbeatRequestState = mock(HeartbeatRequestManager.HeartbeatRequestState.class);
+        errorEventHandler = mock(ErrorEventHandler.class);
+        heartbeatRequestManager = new HeartbeatRequestManager(
+            mockTime,
+            mockLogContext,
+            config,
+            mockCoordinatorRequestManager,
+            mockSubscriptionState,
+            mockMembershipManager,
+            heartbeatRequestState,
+            errorEventHandler);
+    }
+
+    @ParameterizedTest
+    @MethodSource("stateProvider")
+    public void testPoll_sendHeartbeatOnMemberState(final MemberState state) {
```

Comment:
This is not using the state from params so I expect the else part will never be executed (member should be always in the default UNJOINED). Is it missing mocking the state value using the param I guess?

Reply:
you are right! I will rewrite this test just mocking the notInGroup response.